src/System.Web.OData/OData/Query/Expressions/PropertyContainer.cs

```diff
@@ -68,32 +68,39 @@ public static Expression CreatePropertyContainer(IList<NamedPropertyExpression>
             {
                 NamedPropertyExpression property = properties.First();
                 int count = properties.Count - 1;
-                int leftSize = GetLeftSize(count);
-                Expression leftNext = CreatePropertyContainer(properties.Skip(1).Take(leftSize).ToList());
-                Expression rightNext = CreatePropertyContainer(properties.Skip(1 + leftSize).ToList());
-                container = CreateNamedPropertyCreationExpression(property, leftNext, rightNext);
+                List<Expression> nextExpressions = new List<Expression>();
+                int parts = SingleExpandedPropertyTypes.Count - 1;
+                int offset = 0;
+                for (int step = parts; step > 0; step--)
+                {
+                    int leftSize = GetLeftSize(count - offset, step);
+                    nextExpressions.Add(CreatePropertyContainer(properties.Skip(1 + offset).Take(leftSize).ToList()));
```

Comment:
generating new sub list could get expensive. i think you could pass down range(start and end indices) along with the original list

Reply:
Had that conversation with Stan when I created binary-tree implementation.  Doing that logic will require 100% manual implementation and error prone #WontFix